ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/ConnectExecutor.java

```diff
@@ -59,6 +62,11 @@ private ConnectExecutor() { }
       );
     }
 
+    if (createConnector.ifNotExists() && response.httpCode() == HttpStatus.SC_CONFLICT) {
```

Comment:
I think we should check if the connector exists before attempting to call the `client.create`, and only if `ifNotExists` is true. I noticed the `client.create()` code retries up to 3 times in the following cases:
```
.retryIfResult(
              result -> result == null
                  || result.httpCode() >= HttpStatus.SC_INTERNAL_SERVER_ERROR
                  || result.httpCode() == HttpStatus.SC_CONFLICT)
```

There's no need to retry if the connector exists, right? Better just check if it exists to keep this command faster, then return the WarningEntity, otherwise try the create.

Reply:
@big-andy-coates  above had suggested inconsistencies might creep in

> Of course, this would suffer from a race condition: the connector could be added by another system between the connectorExists check and the addConnector call. So really, we also need the addConnector call to gracefully handle the already exists

Would it rather be better to remove `result.httpCode() == HttpStatus.SC_CONFLICT)` from `retryIfResult()` since we will be handling `HttpStatus.SC_CONFLICT` with a WarningEntity , 
Meaning does it make sense to even retry when we know the connector exists

Reply:
Sadly we can't. The `SC_CONFLICT` may also happen if the worker group is in a rebalancing process.

```
POST /connectors
Create a new connector, returning the current connector info if successful. Return 409 (Conflict) if rebalance is in process
```

What do you think of adding the check before attempting to create the connector just to avoid the retries, and also leave the current check you have in this PR in case a race-condition happens? As mentioned by Andy `we also need the addConnector call to gracefully handle the already exists`

Reply:
> SC_CONFLICT may also happen if the worker group is in a rebalancing process.

Ah ! I did not know that ,
so yeah I feel that would be the better option . 
I have updated the PR .. 
**Lots** of conditionals to handle all cases :smile: